Installer: make backend compatible with Terraform 0.10.x

alexsomesan opened the pull request:
This change makes the the backend compatible with Terraform beyond 0.9.x

It adds logic to determine the version of Terraform being used and decides based on version if it needs to run `terraform get` in case of 0.9.x or `terraform init` for 0.10.x and beyond.

Note that the new `init` command also includes the previous functionality of `get` so there is no need to run both.

/cc: @sym3tri

alexsomesan commented:
One test-case failed due to Terraform timeout issues.
IMHO it's fine, we shouldn't need to re-run the whole suite.

sym3tri commented:
I would prefer we re-run tests and only merge once green.

alexsomesan commented:
re-test this please

squat commented on installer/api/handlers_terraform.go:
```diff
@@ -58,12 +60,23 @@ func terraformApplyHandler(w http.ResponseWriter, req *http.Request, ctx *Contex
-	// Execute TerraForm get and wait for it to finish.
-	_, getDone, err := ex.Execute("get", "-no-color", tfMainDir)
+	// Choose to run 'get' or 'init' based on Terraform version.
+	verRx := regexp.MustCompile("^Terraform v0\\.[0-9]\\.[0-9]+")
```
nit: this regular expression only checks if a string is version <0.10, not if it is a valid version in general, so consider renaming for clarity and maintainability, e.g. `sub10Rx`

alexsomesan replied:
Don't have any preference here. I'll rename it.

squat commented on installer/api/handlers_terraform.go:
```diff
@@ -58,12 +60,23 @@ func terraformApplyHandler(w http.ResponseWriter, req *http.Request, ctx *Contex
+		prepCommand = "get"
+	}
+
+	// Execute TerraForm get or init and wait for it to finish.
```
s/TerraForm/Terraform/

alexsomesan replied:
Good spot!
This is all over the file, and thus got copy-pasted in here.
I'll fix it everywhere.

squat commented on installer/api/handlers_terraform.go:
```diff
@@ -58,12 +60,23 @@ func terraformApplyHandler(w http.ResponseWriter, req *http.Request, ctx *Contex
+	// Execute TerraForm get or init and wait for it to finish.
+	_, prepDone, err := ex.Execute(prepCommand, "-no-color", tfMainDir)
+	if err != nil {
+		return newInternalServerError("Failed to run TerraForm (%s): %s", prepCommand, err)
```
s/TerraForm/Terraform/

alexsomesan replied:
copy-paste. will fix.

squat commented on installer/api/handlers_terraform.go:
```diff
@@ -58,12 +60,23 @@ func terraformApplyHandler(w http.ResponseWriter, req *http.Request, ctx *Contex
+		return newInternalServerError("Failed to determine Terraform version: %s", err)
 	}
-	<-getDone
+	var prepCommand = "init"
```
typically, we don't use `var` within func blocks. Consider rewriting as `prepCommand := "init"`

alexsomesan replied:
Make sense. I'll adapt it.

alexsomesan commented:
@squat addressed all the nits.

Looks like the 'TerraForm' spelling is all over the place, also in other areas of the code. I'll follow-up with another PR to align it, just to keep this one concise.

squat commented:
All tests passed except for one of the AWS tests, which timed out during the cluster integrity tests. No need to re-run tests IMO. Go ahead and merge it.

cpanato commented:
Agree, the test that failed passed before in another test execution.
